### ISSUE TITLE
Customizing error messages on failed check constraints

### ISSUE THREAD

**Issue by steve-chavez** [PostgREST/postgrest#2706]
### Problem

Having:

```sql
CREATE DOMAIN public.filled_text AS text
	CONSTRAINT not_empty CHECK ((NOT (VALUE ~ '^\s*$'::text)));

COMMENT ON DOMAIN public.filled_text IS 'Text value that cannot be empty or whitespace, but doesn''t assume not null.';
```

A failed POST request will return:

```json
{"code":"23514","details":null,"hint":null,"message":"value for domain filled_text violates check constraint \"not_empty\""}
```

Which is not user friendly. We've been telling users to format the error message on the client side, but this is not possible with direct consumers of the API(unless a proxy is involved, which complicates the setup).

### Proposal

Just to kickstart discussion. We could have a `pre-request` function that does:

```sql
create or replace function custom_errors() returns void as $$
declare
  error_message text := current_setting('response.error', true)::json->>'message';
  error_code text := current_setting('response.error', true)::json->>'code';
begin
  if error_code = 23514 and error_message like '%not_empty%'
    then set_config(`response.error`, json_build_object('message', 'cannot be empty'); -- one problem here is that there's no way to know the column
  end if;
end; $$ language plpgsql;
```

**Comment by mkleczek** [comment]
Maybe https://datatracker.ietf.org/doc/rfc7807/ is relevant?

**Comment by steve-chavez** [comment]
Hm, how so? RFC 7807 is for the error message format right? We discussed adopting that on https://github.com/PostgREST/postgrest/pull/1926#issuecomment-916591592

**Comment by steve-chavez** [comment]
On https://github.com/PostgREST/postgrest/issues/3084#issue-2022449376, @dvv mentioned:

> I wonder if we could have a post-request db hook in addition to pre-request one?
The rationale is to sanitize errors or to reconcile just in-place.
It could be given all available info from get stacked diagnostics and raise more meaningful messages or just rewrite the json data.

Hm, I don't think we can use [get stacked diagnostics](https://www.postgresql.org/docs/current/plpgsql-control-structures.html#PLPGSQL-EXCEPTION-DIAGNOSTICS) since that would imply `plpgsql`?

**Comment by dvv** [comment]
@steve-chavez
Right, catching the exception directly implies plpgsql `begin/exception`.

But I don't know what is the complete set of fields available in `SQL.ServerError` at https://github.com/PostgREST/postgrest/blob/main/src/PostgREST/Error.hs#L433-L437

If they contain `schema_name, table_name, column_name, constraint_name, stack` we could just pass the entire `SQL.ServerError` thru our `db-error-handler` hook function.

**Comment by steve-chavez** [comment]
When doing this, let's make sure the scenarios mentioned on https://github.com/PostgREST/postgrest/issues/3980#issuecomment-2766043935 are covered.
